nucypher/policy/reservoir.py

```diff
@@ -0,0 +1,115 @@
+"""
+This file is part of nucypher.
+
+nucypher is free software: you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+nucypher is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+from typing import Iterable, Optional, List
+
+from eth_typing import ChecksumAddress
+
+from nucypher.blockchain.eth.agents import StakersReservoir, StakingEscrowAgent
+
+
+def make_federated_staker_reservoir(learner: 'Learner',
+                                    exclude_addresses: List[str] = None,
```

Comment:
Would `ChecksumAddress` fit better here instead of `str`? And, perhaps, `Optional[Iterable]` instead of `List`.

Reply:
Or you could make `()` the default value, thus avoiding the need for `Optional`, and creating `exclusion_set` just as `exclusion_set = set(include_addresses) | set(exclude_addresses)`

Reply:
Not a fan of the default empty set, but made the param Optional[Iterable] and used ChecksumAddress - ✔️

Reply:
Yes, I know that mutable defaults is a controversial thing, but I was talking about a `()`, an empty tuple. Should be fine, it's immutable.

Reply:
I rather ensure that the `None` case is handled by accepting `None` as the default, as opposed to using a tuple as the default, and `None` could be passed by callers and not handled appropriately. Using a tuple as default forces extra code in callers of the function eg. `Porter.get_ursulas()` to prevent `None` from being passed to the reservoir function.

I did some cleanup using the recommendation about set intersection though 👍 .